### DISCUSSION TITLE
When I asynchronize a function I get the error: Objects are not valid as a child of React (found: [Promise object]).

### DISCUSSION THREAD

**Discussion by MailenRosario** [community/community#71918] [Programming Help]
### Body

I have a component that receives an array with objects. These objects can have a type and depending on this, they should render different components.

It happens to me that the type: reviews needs to make some checks and depending on these, calls to different APIs.

Then as I have to work with asynchronous data, I need the function to be asynchronous. I want to be able to make, inside my renderComponents function, in the case "reviews": of the switch, a const data = await getReviews(element); and for this I need the renderComponents function to be asynchronous.

The issue is that when I add async I get the error:

Error: Objects are not valid as a React child (found: [object Promise]). If you meant to render a collection of children, use an array instead.

I get the error even commenting the case "reviews", when I add the word async to the function. If I remove the word async everything works fine.

I tried doing this type of function and also arrow functions, but the error persists.

```javascript
async function renderComponents(element) {
    const type = element.sys?.contentType?.sys?.id || null;
    if (type) {
      switch (type) {
        case "buttonOrLink":
          return <p>Component for buttonOrLink</p>;
          
        case "copyImagegridSection":
          return <p>Component for copyImagegridSection</p>;

        case "explore":
          return <p>Component for explore</p>;

        case "video":
          return (
            <LightBox
              content={element.fields.videoThumbnail?.fields?.file?.url}
              video={element.fields}
            />
          );

        // case "reviews":
          // const data = await getReviews(element);
          // console.log(data);
          // break;

        default:
          return;
      }
    }
    return;
  }

  return (
    <>
      Page Section
      {content.map(el => {
        return (
          <div key={el.sys.id}>
            {renderComponents(el)}
          </div>
          
        );
      })}
      {offers.length > 0 && <Specials content={offers} />}
      {faqs.length > 0 && <FAQ faqs={faqs} />}
    </>
  );
```

EDIT

One thing that happens to me, that I don't understand but maybe it can be useful for someone to notice, is that if I uncomment this block

```javascript
// case "reviews":
// const data = getReviews(element);
// console.log(data);
// break;
```

Without making the renderComponents function asynchronous, and placing a console.log in the getReviews() function, when the page is loaded, first prints the console.log(data) with Promise {<pending>} and then prints the getReviews console showing the correct data.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by KernelDiego** [comment]
Problem
The problem is that React expects the component to return an element or an array of elements, not a promise or asynchronous functions. In your case, the renderComponents function should return a synchronous React element, and then handle the asynchronous logic within that component.

Solution approach
One way to do this is to use the component state to store the result of the promise and update it once it is resolved. If we use the useEffect hook to handle the logic of getting the data, then, in the Reviews component, you can have an empty initial state for the data, and use use useEffect to make the API call and update the state with the returned data.



```javascript
const Reviews = ({ element }) => {
    const [reviews, setReviews] = useState([]);

    useEffect(() => {
        getReviews(element).then(data => setReviews(data));
    }, [element]);

    if (!reviews) {
        return <p>Cargando...</p>;
    }

    // Acá tocaría renderizar componente con los datos
    return <div>...</div>;
};
```

Then, the renderComponents function would look something like this:


```javascript
function renderComponents(element) {
    const type = element.sys?.contentType?.sys?.id || null;
    if (type) {
        switch (type) {
            ...
            case "reviews":
                return <Reviews element={element} />;
            ...
```

**Comment by queenofcorgis** [comment]
GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this post. Please note any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
